Starter page templates - fix double modal / wrong button appearing - alt approach

Addison-Stavlo opened the pull request:
#### Changes proposed in this Pull Request

Alternative approach to #48377 to solve #48370.

Behavior, description, and testing instructions should be the same as #48377.  Here we remove the separate instance of the modal and instead trigger the original one with an extra store state value to determine its context.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

Fixes #48370

Addison-Stavlo committed:
remove extra use of component, add sidebar state to store

Addison-Stavlo committed:
remove unnecessary return nesting

p-jackson reviewed (commented):
Yes I do prefer this approach over #48377! Now it feels like the modal component is fully controlled by the data store rather than half by the store and half by some components.

Another thought I had is that rather than juggling two booleans, the state is better encoded like an enum? i.e. the type of the state could be:
```typescript
interface State {
	openState: 'CLOSED' | 'OPEN_FROM_SIDEBAR' | 'OPEN_FROM_ADD_PAGE';
}
```

The only thing it really gives us is it makes this impossible state: `{ isOpen: false, isPromptedFromSidebar: true }`, actually impossible.

We'd probably still end up wanting seperate `isOpen` and `isPromptedFromSidebar` selectors though, so maybe it's not worth it. I'm only 60:40 convinced it's a good idea.

p-jackson commented:
We can remove the `templatesPluginSharedProps` props being passed [here](https://github.com/Automattic/wp-calypso/blob/044abff1a06f3630be32cbfa50d7456bf348293e/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/index.js#L62) now.

Addison-Stavlo committed:
update store to enum, fix existing pages bug

Addison-Stavlo commented:
> I'm only 60:40 convinced it's a good idea.

It seems like a good idea to me!  Tried it out in db2b0dd:

* Refactored store to use the enum
* Removed the unnecessary props noted above.
* Moved `registerPlugin` out of the screenAction condition.  It turns out the original plugin for the new page side was only registered conditionally, so this branch introduced a bug where we couldn't actually open the modal for existing pages through the sidebar.

Addison-Stavlo committed:
minor comment

Addison-Stavlo committed:
consolidate plugin registration

Addison-Stavlo commented:
Also realized theres no point to registering 2 separate plugins at this point, so consolidated them into one. e45d84d

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/index.js:
```diff
@@ -41,40 +41,34 @@ const templatesPluginSharedProps = {
+				<PluginDocumentSettingPanel
+					name="Template Modal Opener"
+					title={ __( 'Page Layout', 'full-site-editing' ) }
+					className="page-template-modal__sidebar" // eslint-disable-line wpcalypso/jsx-classname-namespace
```
A tidy up thing, this eslint comment is no longer needed as this warning's now disabled for all of ETK in `editing-toolkit/.eslintrc.js`

p-jackson reviewed (approved):
This is great! Having just the single instance of the modal and registering just the one plugin has tidied up a bunch of the unnecessary complexity for me. The "was opened from sidebar" stuff is now just about changing the tracking prop and the close button behaviour, which is all I would have expected. Nice.

Addison-Stavlo committed:
remove eslint comment

joanrho commented:
Hey, @Addison-Stavlo! Appreciate the prompt response. I was just testing and noticed that the current behavior didn't match the proposed fixes mentioned in [#48370](https://github.com/Automattic/wp-calypso/issues/48370):

> **Proposed changes**
> * “Recent posts” sidebar link from within the Editor > "Add new page" > new page layout modal with close icon that returns users to the Editor on the post/page they were on.
> * Customer Home Quick link > "Add a page" > new page layout modal with close icon that return users to Customer Home.
> * WP.com/pages > "Add new page" > new page layout modal with close icon that returns users to WP.com/pages.

Specifically, I'm still seeing the "back" `<` icon across all three entry points mentioned above, instead of the correct "close" `x` icon. Also, when I enter the new page layout modal from the “Recent posts” sidebar link from within the Editor, exiting it via the close icon should return me back to the Editor page/post I was in (not Calypso /pages).

Addison-Stavlo commented:
> I'm still seeing the "back" < icon across all three entry points mentioned above

Oh I didn't realize we were changing the icons for the given states, we just fixed it so that the expected states show up in the right circumstances.  

What I mean is, this was originally designed to appear as either a "<" back icon leading back out of the editor, or a "x" close icon closing the modal so the user could continue to use the editor.  "<" was always supposed to be present from adding a new page, and "x" when opening from the sidebar.  The bug this fixed was when adding a new page, both modals would open with the "x" version being on top and visible, meaning if the button was pressed it would close the modal and go into the editor as opposed to leading back as would be expected if the proper "<" version showed up as originally intended.  After that, if one were to open it from the sidebar, it would toggle both versions again with the "<" on top this time causing it to leave the editor instead of close the modal as would happen if the expected "x" version showed up.  So reiterating, "<" was originally intended to signify leaving the editor and "x" as closing the modal to continue using the editor.

We can update to change the icons as well if that is desired.  But if we are changing "<" to be "x" to signify leaving the editor, what do we then replace "x" with to signify just closing the modal when its opened from the sidebar?  

My apologies for misunderstanding, when I read 'close icon that returns user to (outside of the editor)' in the issue I was assuming "<" since that was the icon set up to do that leaving the editor action in the original design.

> Also, when I enter the new page layout modal from the “Recent posts” sidebar link from within the Editor, exiting it via the close icon should return me back to the Editor page/post I was in (not Calypso /pages).

Ah yeah we were discussing this earlier.  We will need to make a new issue for this one.  We can find a solution for that, but its currently set up to lead to the same place as all other buttons leaving the editor and in this special case we probably wouldn't want those other editor exit buttons to lead back to the previous editor screen.

joanrho commented:
@Addison-Stavlo Thanks for the info around the intentions behind the back vs close icon usage! 

I don't think it's clear that "x" signals that users will be leaving the editor rather than just closing/dismissing the modal, which is why I proposed we use the close icon for all three cases. The expected destinations should be where the users came from: from Editor, back to Editor; from Calypso Pages, back to Calypso Pages; from Customer Home Quick link, back to Customer Home.

Addison-Stavlo commented:
> Thanks for the info around the intentions behind the back vs close icon usage!

No problem, its just the historical usage I found in the code but nothing we are married to.  "<" was the only icon when SPTs were only triggered from the new page flow, then we added the option to open it from the sidebar and used "x" for that version instead.

> I don't think it's clear that "x" signals that users will be leaving the editor rather than just closing/dismissing the modal, which is why I proposed we use the close icon for all three cases. 

So if I am understanding correctly, we should have the same icon in all cases.  And that should be the "x" (no "<" anywhere) ?  

My main concern here is conflating the same icon with different expectations and action types:

Case A - 2 icons - A user has one icon "<" from the 'new page' flow that always navigates back out of that instance of the editor to where they came from.  When a user opens the same familiar modal from the sidebar and sees a different icon than they are used to ("x" currently) there is no mistaking that this does a different action - it does not navigate back out of the editor.  They can freely dismiss the modal without a second thought.

Case B - 1 icon - However, if a user sees the same icon in the sidebar flow they could immediately associate it with the same action they are used to for that icon - navigating back out of that instance of the editor.  They could then be reluctant to use this button to close the modal, as they want to remain in the editor and don't want to navigate out of it.  From there they could have problems figuring out how to close the modal, and accidentally select the 'blank' layout mistaking it as they way to get out of there without adding anything new to their page.  This confusion could result in more user issues and calls to HEs for help after accidentally overwriting content, etc.

I think there is definitely some value and clarity in having separate icons since the actions do carry different significance.  While they both 'go back' generally speaking, some do and some do not leave that instance of the editor which seems to be an important distinction.
